### ISSUE TITLE
Write Timeouts during rebuild

### ISSUE THREAD

**Issue by juliayakovlev** [scylladb/scylladb#9053]
The original issue was changed in order to capture the later issue that is being discussed below.
The original issue root cause was the commit log hard limit, but this one has several duplicates already.

This issue should focus on the c-s failure with WriteTimeoutException (same symptoms as the original issue), but only on the special case of the nemesis RebuildStreamingErr (HWLB and cache invalidation are the current suspects).

The relevant comments for this issue starts at https://github.com/scylladb/scylla/issues/9053#issuecomment-903279392.

**Comment by juliayakovlev** [comment]
https://github.com/scylladb/scylla/issues/9054#issuecomment-882049557

**Comment by bhalevy** [comment]
@asias please look into this.

**Comment by asias** [comment]
No throttle is applied to the c-s workload during the test.  The system is loaded 100% on most of the shards. The cluster can not keep up with the workload (c-s + repair) so some of the writes time out which I think is expected. 

![image](https://user-images.githubusercontent.com/320076/127104139-e1590dc4-d987-4dfd-831e-ed5b65538466.png)

![image](https://user-images.githubusercontent.com/320076/127104212-d4c53c48-6e29-4209-b5c5-9ceb45a78af5.png)


```
prepare_write_cmd: ["cassandra-stress write cl=QUORUM n=100000000 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=100 -pop seq=1..100000000"]

stress_cmd: ["cassandra-stress mixed cl=QUORUM duration=4320m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native  -rate threads=100 -pop seq=1..100000000 -log interval=5",
             "cassandra-stress user profile=/tmp/cs_mv_profile.yaml ops'(insert=3,read1=1,read2=1,read3=1)' cl=QUORUM duration=4320m -port jmx=6868 -mode cql3 native -rate threads=20"]

stress_read_cmd: ["cassandra-stress read cl=QUORUM duration=4320m -port jmx=6868 -mode cql3 native  -rate threads=50 -pop seq=1..100000000 -log interval=5"]

run_fullscan: 'keyspace1.standard1, 60' # 'ks.cf|random, interval(min)''

n_db_nodes: 6
n_loaders: 3
n_monitor_nodes: 1
```

**Comment by roydahan** [comment]
After some bisecting work done with the help of @bhalevy and @slivne, this is the inforamtion we were able to collect:
- On 4.4.0 and 4.5 without the commitlog hard limit commits the issue doesn't reproduce.
- The issue reproduces regardless of MgmtRepair, Full Scan and/or any other nemesis running (reproduced with NOOPMonkey).
- This test  workload is overriding the same dataset (50gb) for long period of time (maybe related to a vulnerability of the hard limit may have).

(I chose this issue to add the information because the information I have below, is from a run with "MgmtRepair").

Live Monitor: (time range is set where the issue reproduced - 05/08/21 11:47 UTC)
http://100.27.7.58:3000/d/t8g0gDM7z/longevity-50gb-3d-reproducer-with-rpms-scylla-per-server-metrics-nemesis-master?orgId=1&from=1628145421175&to=1628171409833

"Live Snapshot": 
https://snapshot.raintank.io/dashboard/snapshot/ccw1yZKuIokL5Y0wXrXzxyLlrYFOrWZj
https://snapshot.raintank.io/dashboard/snapshot/C5sdS70Aup34NIxbUgCtna0cPymYoShf
https://snapshot.raintank.io/dashboard/snapshot/IoExWf1Urq9LVRAvf7Ygja5xmC5FrOkG

**Comment by bhalevy** [comment]
@elcallio so our investigations point at the commitlog hard limit patches.  Other than reverting them in 4 5 until the issue is fixed, can we bypass the logic by default using the configuration (like the hints manager does)?

**Comment by elcallio** [comment]
The `commitlog::config` option is `allow_going_over_size_limit`, but it is not exposed through global config. 
I would _really_ (again) suggest running the test with log level `commitlog=debug` and see if you actually get hangs on segment alloc. 
Shlomi requested/suggested adding a few counter that expose segment alloc/wait as well, to be able to view this through metrics. I can send a patch with this.

**Comment by elcallio** [comment]
Actually, the commit log counter `pending_allocations`, when viewed against disk size and flush count should give a fair view iff the commitlog is stalled as well.

**Comment by roydahan** [comment]
@elcallio what's the exact query you run to check that?
I have several more monitors I can check them as well.

I'm running now a test with commitlog debug level.

**Comment by elcallio** [comment]
@roydahan I guess just turn on all commitlog counters in the metrics view. As I said, `pending_allocations`, `segments`, `disk_total_bytes` and the derivative of `flush` might be most interesting. 

I just realized that making a patch against master adding new counters to explicitly track segment alloc stalls might be not be so great, since master i coroutinized now (so won't apply to 4.5). I can make a 4.5 patch if you want?

**Comment by roydahan** [comment]
Patch for 4.5 is also fine.

**Comment by roydahan** [comment]
http://100.27.15.184:3000/d/dRdTwlG7k/longevity-50gb-3d-reproducer-scylla-per-server-metrics-nemesis-master?orgId=1&from=1628596800000&to=1628614800000

**Comment by elcallio** [comment]
Thanks @roydahan .
A quick look at the logs shows that, yes, it looks like we're getting stuck on segment allocation. We get quite a few stalls where we exceed the disk usage limit. _But_ we are _not_ getting stuck. It is just memtable flushes that seem to take time visa-vi the data throughput. If I grep out just commitlog stuff for a single shard, we have a distinct pattern where we fill up the commitlog space, issue flush requests into database tables, then have loads and loads of callbacks that however still do not clear all the usage markers, meaning that we can't fully release any files. Eventually we _do_ get them cleared, and things start rolling again, but it takes a while. Up to 20s.
Either there is a regression in flush handling, or this is an old problem that was sort of hidden before by unbounded commitlog size increase, that we have long periods where things are not yet in sstable. 
I can see a few obvious approaches: 
1.) Disable the size enforcement. Just flip the default to "true" for allow size overflow in commitlog config. 
2.) Tweak default commitlog size - maybe the default is too little? But then we need to tweak the boundaries for when to start flushing as well.

**Comment by elcallio** [comment]
Attached grep:ed log.
[cl0.txt](https://github.com/scylladb/scylla/files/6968139/cl0.txt)

Last pattern occurrance:
```
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Disk usage (8160 MB) exceeds maximum (8154 MB) - allocation will wait...
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Flushing (0 MB) to {0, 446945, 0}
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Discarding b4dbb7b4-dc49-3fb5-b3bf-ce6e434832ca: {{ 446941 -> 1}, { 446699 -> 1}, { 446703 -> 1}, { 446716 -> 1}, { 446715 -> 1}, { 446721 -> 2}, { 446440 -> 1}, { 446817 -> 1}, { 446759 -> 1}, { 446784 -> 1}, { 446903 -> 1}, { 446729 -> 1}, { 446787 -> 1}, { 446852 -> 1}, { 446899 -> 1}}
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Not safe to delete segment CommitLog-2-446439.log; dirty is 33393d50-f9dd-11eb-9f6a-5daff5a9a27232386f20-f9dd-11eb-a649-7c6ab3a29a843060c8f0-f9dd-11eb-b187-36f740c5d611
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Not safe to delete segment CommitLog-2-446440.log; dirty is 32386f20-f9dd-11eb-a649-7c6ab3a29a843060c8f0-f9dd-11eb-b187-36f740c5d61133393d50-f9dd-11eb-9f6a-5daff5a9a272
Aug 11 06:15:34 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Not safe to delete segment CommitLog-2-446444.log; dirty is 32386f20-f9dd-11eb-a649-7c6ab3a29a843060c8f0-f9dd-11eb-b187-36f740c5d61133393d50-f9dd-11eb-9f6a-5daff5a9a272
.....
Aug 11 06:15:52 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Discarding 33393d50-f9dd-11eb-9f6a-5daff5a9a272: {{ 446945 -> 4957}, { 446944 -> 12821}, { 446943 -> 12789}, { 446942 -> 12870}, { 446941 -> 12848}, { 446940 -> 13020}, { 446939 -> 12981}, { 446938 -> 12717}, { 446937 -> 12801}, { 446936 -> 12764}, { 446935 -> 12861}, { 446934 -> 
...
Aug 11 06:15:52 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Segment CommitLog-2-446709.log is no longer active and will submitted for delete now
...
Aug 11 06:15:52 longevity-tls-50gb-3d-repro-wi-db-node-635b8953-2 scylla[14318]:  [shard 0] commitlog - Using recycled segment file /var/lib/scylla/commitlog/Recycled-CommitLog-2-446948.log -> /var/lib/scylla/commitlog/CommitLog-2-446950.log

**Comment by bhalevy** [comment]
Or we should use an adaptive approach where we start flushing earlier and flush more aggressively as the disk usage grows.

**Comment by bhalevy** [comment]
> I can see a few obvious approaches:
> 1.) Disable the size enforcement. Just flip the default to "true" for allow size overflow in commitlog config.
> 2.) Tweak default commitlog size - maybe the default is too little? But then we need to tweak the boundaries for when to start flushing as well.

@elcallio, @slivne, @roy
how about the following game plan:
1. Disable the size enforcement - try first with SCT on 4.5 or master and if it works submit patch to master and backport to 4.5
2. Work on the real fix, e.g. soft/hard limit or adaptive flush intensity and when fixed in master, consider backport to 4.5

**Comment by elcallio** [comment]
I am all for it. Do you want me to send a patch for disabling?

**Comment by elcallio** [comment]
Also, if we flip this, should I add a global config flag to turn on commitlog hard limit?

**Comment by bhalevy** [comment]
> I am all for it. Do you want me to send a patch for disabling?

yes please, and if possible produce dist packages for @roydahan 

> Also, if we flip this, should I add a global config flag to turn on commitlog hard limit?

IMHO it's a good idea.  I think we need a safety net out in the field until this is proven to be stable.

**Comment by elcallio** [comment]
rpms: gs://scratch.scylladb.com/calle/9053

**Comment by roydahan** [comment]
I'm on it

**Comment by roydahan** [comment]
After @bhalevy built ubuntu packages (one of the rpms was broken), I finally was able to run the test.
unfortunately, I lost the results from the weekend due to enospc on the builder, so had to start it again.
The test is now running for more than 12h and looks good (previously it was easily reproduced after 4-5 hours).

http://44.192.109.155:3000/d/jPPQAN7nk/longevity-50gb-3d-reproducer-scylla-per-server-metrics-nemesis-master?orgId=1&refresh=30s&from=now-12h&to=now

**Comment by bhalevy** [comment]
Sounds encouraging, once commit "b0edfa6d7 commitlog/config: Make hard size
enforcement false by default + add config opt"
is promoted to branch-4.5 we can build and test 4.5.rc6 re-run and
hopefully get 4.5.0 out of the door.

**Comment by roydahan** [comment]
Test with 4.5.rc6-0.20210817.88314fedf (ami-0c8070384fb84cdb3) passed successfully 3 days run of the reproducer.

**Comment by juliayakovlev** [comment]
The test `longevity-50gb-3days-test` running version `4.5.rc6-0.20210817.88314fedf` failed with cassandra-stress WriteTimeoutException.
Timeouts started during repair_streaming_err nemesis, but c-s failed more then one hour later, not during nemesis.\

During nemesis repair was started and interrupted (by hard reboot of the instance). After the instance was back ,
rebuild was started:
```
< t:2021-08-22 12:51:07,289 f:remote_base.py  l:520  c:RemoteCmdRunner      p:DEBUG > Running command "/usr/bin/nodetool -u cassandra -pw cassandra  rebuild "...
```

Timeout were started at `2021-08-22 13:17:25,132`:
```
< t:2021-08-22 13:17:25,175 f:base.py         l:220  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
< t:2021-08-22 13:17:30,616 f:cluster.py      l:1531 c:sdcm.cluster         p:DEBUG > 2021-08-22T13:17:24+00:00  longevity-tls-50gb-3d-4-5-db-node-8535a63b-4 !INFO    | scylla:  [shard 0] range_streamer - Rebuild with 10.0.2.115 for keyspace=keyspace1, streaming [144, 160) out of 160 ranges
< t:2021-08-22 13:18:11,729 f:base.py         l:220  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.ReadTimeoutException: Cassandra timeout during read query at consistency QUORUM (2 responses were required but only 1 replica responded)
```

green line - node were rebuild is running:

![Screenshot from 2021-08-22 17-33-39](https://user-images.githubusercontent.com/34435448/130358971-36047b0e-bd11-4a32-9667-568c5f98cb79.png)

Nemesis finished, but timeouts continued

There are a lot of `seastar::semaphore_timed_out`:
```
< t:2021-08-22 13:28:21,189 f:cluster.py      l:1531 c:sdcm.cluster         p:DEBUG > 2021-08-22T13:28:07+00:00  longevity-tls-50gb-3d-4-5-db-node-8535a63b-4 !WARNING | scylla:  [shard 4] storage_proxy - Failed to apply mutation from 10.0.0.240#4: seastar::semaphore_timed_out (Semaphore timedout)
< t:2021-08-22 13:28:21,190 f:cluster.py      l:1531 c:sdcm.cluster         p:DEBUG > 2021-08-22T13:28:07+00:00  longevity-tls-50gb-3d-4-5-db-node-8535a63b-4 !WARNING | scylla: message repeated 9 times: [  [shard 4] storage_proxy - Failed to apply mutation from 10.0.0.240#4: seastar::semaphore_timed_out (Semaphore timedout)]
```

Cassandra-stress failed at `2021-08-22 13:38:55.184`
```
2021-08-22 13:38:55.184: (CassandraStressLogEvent Severity.CRITICAL): type=OperationOnKey regex=Operation x10 on key\(s\) \[ line_number=38921 node=Node longevity-tls-50gb-3d-4-5-loader-node-8535a63b-3 [13.49.245.26 | 10.0.0.4] (seed: False)
java.io.IOException: Operation x10 on key(s) [4f3730394b4f36314e30]: Error executing: (WriteTimeoutException): Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
```

[Live monitor](http://13.51.156.156:3000/d/RbV85wn7z/longevity-50gb-3days-test-scylla-per-server-metrics-nemesis-master?orgId=1&from=1629637895062&to=1629639564661&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=CassandraStressEvent)

*Installation details*
Kernel version: `5.4.0-1035-aws`
Scylla version (or git commit hash): `4.5.rc6-0.20210817.88314fedf`
Cluster size: 6 nodes (i3.4xlarge)
Scylla running with shards number (live nodes):
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-1 (13.53.35.126 | 10.0.2.115): 14 shards
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-2 (13.53.125.104 | 10.0.1.226): 14 shards
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-3 (13.49.77.102 | 10.0.2.26): 14 shards
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-4 (13.51.235.202 | 10.0.0.113): 14 shards
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-5 (13.49.226.92 | 10.0.3.158): 14 shards
   longevity-tls-50gb-3d-4-5-db-node-8535a63b-6 (13.51.55.91 | 10.0.0.240): 14 shards
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-074912d764ff480e5` (aws: eu-north-1)


Test: `longevity-50gb-3days-test`
Test name: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):
- [longevity-50GB-3days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/a192e58bf41a53b06cb23c3eb83926943583f13d/test-cases/longevity/longevity-50GB-3days-authorization-and-tls-ssl.yaml)

Restore Monitor Stack command: `$ hydra investigate show-monitor 8535a63b-fa03-4b1f-b533-185a78247d82`
Show all stored logs command: `$ hydra investigate show-logs 8535a63b-fa03-4b1f-b533-185a78247d82`

Test id: `8535a63b-fa03-4b1f-b533-185a78247d82`


Logs:

db-cluster - [https://cloudius-jenkins-test.s3.amazonaws.com/8535a63b-fa03-4b1f-b533-185a78247d82/20210822_134458/db-cluster-8535a63b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8535a63b-fa03-4b1f-b533-185a78247d82/20210822_134458/db-cluster-8535a63b.tar.gz)
sct-runner - [https://cloudius-jenkins-test.s3.amazonaws.com/8535a63b-fa03-4b1f-b533-185a78247d82/20210822_134458/sct-runner-8535a63b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8535a63b-fa03-4b1f-b533-185a78247d82/20210822_134458/sct-runner-8535a63b.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-4.5/job/longevity/job/longevity-50gb-3days-test/6/)

**Comment by bhalevy** [comment]
I'm not sure if that's related or not but there are a lot of deletes on this node when the timeouts happen:

![Screenshot from 2021-08-23 13-36-26](https://user-images.githubusercontent.com/20910904/130433914-ee4b5dc5-8c75-4feb-a0b6-77aac9d08bd9.png)
![Screenshot from 2021-08-23 13-39-14](https://user-images.githubusercontent.com/20910904/130434183-cf833fc0-93a5-4162-8aa9-9bc8f5b131cf.png)

**Comment by slivne** [comment]
We can check the LSA memory usage (per shard) and see why

**Comment by elcallio** [comment]
Do we have any test setup that can provoke these timeouts/stalls in a smaller setting/shorter time span? I've tried provoking it using the same stress-commands as in OP config + small CCM cluster. I can get small stalls, but that pretty much requires me to set commit log sizes to ridiculously small sizes, and even then I have very short stalls, mostly in the order of 10-100ms. 
I assume physical node constraints and the insistent hammering of the nemsis (with subsequent repairs and whatnot) is the actual culprit in stealing disk/cpu bandwidth -> stalls. That again would suggest we need more "emergency" headroom in CL when we hit such a rough patch.

**Comment by roydahan** [comment]
@elcallio unfortunately, we don't have such test  setup.
We couldn't find an easier or shorter way to do it (but we didn't invest much on it as it was easily reproduces on 4.5.rc5 with what we had).

@juliayakovlev is running two more tests:
1. With commitlog set to debug level - running already 24h and show no sign of the issue.
2. Re-running the original test - running about 2 hours.

**Comment by slivne** [comment]
I am not able to see any issue in commitlog with 4.5.rc6 - e.g. there are no writes blocked on memtable / commitlog - which based on what Calle shared should happen. So we need to split this apart.

With regards to this problem the problem is not with commitlog AFAICT in 4.5.rc6 something else maybe at fault. 

(BTW I am not able to access the monitoring any more) - @bhalevy / @asias we need you guys to look at this issue now with 4.5.rc6 - after disabling commitlog hard limit - the problem seems to be different then it was before.

**Comment by juliayakovlev** [comment]
> I am not able to see any issue in commitlog with 4.5.rc6 - e.g. there are no writes blocked on memtable / commitlog - which based on what Calle shared should happen. So we need to split this apart.
> 
> With regards to this problem the problem is not with commitlog AFAICT in 4.5.rc6 something else maybe at fault.
> 
> (BTW I am not able to access the monitoring any more) - @bhalevy / @asias we need you guys to look at this issue now with 4.5.rc6 - after disabling commitlog hard limit - the problem seems to be different then it was before.


I am running now two tests:
1. With commitlog set to debug level (reproducer) - https://jenkins.scylladb.com/job/scylla-4.5/job/Reproducers/job/longevity-50gb-3days-commitlog-debug/3/
2. Re-running the original test (without commitlog debug level) - https://jenkins.scylladb.com/job/scylla-4.5/job/longevity/view/Manually%20Triggered%20Jobs/job/longevity-50gb-3days-test/lastBuild/

Official job (#2) failed with "Operation x10 on key" during RepairStreamingErr nemesis (again). 
[Monitor](http://13.48.48.0:3000/d/dMrIT3nnz/longevity-50gb-3days-test-scylla-per-server-metrics-nemesis-master?orgId=1&from=1629867327038&to=1629873291195&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=CassandraStressEvent)

Reproducer (#1) is continue to run despite the RepairStreamingErr was run 4 times ([Monitor](http://13.51.238.246:3000/d/Mlvkru77z/longevity-50gb-3days-commitlog-debug-scylla-per-server-metrics-nemesis-master?orgId=1&from=1629709023000&to=now&refresh=30s)).

RepairStreamingErr nemesis was run twice during official job. First time it succeeded, second time failed

**Comment by bhalevy** [comment]
@juliayakovlev please fix the hyperlink to jenkins not point to `lastBuild` since this is dynamic and changes over time.  What was the build number? 7?

**Comment by slivne** [comment]
@elcallio can you please look at this time range and see if you cna detect anything that shows commitlog maybe at blame

http://13.48.48.0:3000/d/detailed-4-5/detailed?orgId=1&from=1629867327000&to=1629874527000

(there are no commitlog debug level info here - so its probably harder - but please try)

**Comment by slivne** [comment]
We are going to revert the whole code of commitlog hardlimit fro 4.5 branch and spin 4.5.rc7 and see the result.

**Comment by elcallio** [comment]
> 
> 
> @elcallio can you please look at this time range and see if you cna detect anything that shows commitlog maybe at blame
> 
> http://13.48.48.0:3000/d/detailed-4-5/detailed?orgId=1&from=1629867327000&to=1629874527000
> 
> (there are no commitlog debug level info here - so its probably harder - but please try)

I am as usual probably reading this wrong, but from what I can see, both `pending_allocations` and `requests_blocked_memory` seem to be zero? If so, nothing is actually blocked on commit log. Or even doing commitlog alloc at this time, i.e. no writes are being done...?

**Comment by bentsi** [comment]
@nyh @gleb-cloudius live Graphana: http://13.48.48.0:3000/d/detailed-4-5/detailed?orgId=1&from=1629790541641&to=1629877019596

**Comment by avikivity** [comment]
Looking at a narrower range:

http://13.48.48.0:3000/d/detailed-4-5/detailed?orgId=1&from=1629858550252&to=1629873394771

**Comment by avikivity** [comment]
Reciprocal miss rate climbs to 100 on all nodes (indicating perfect cache hits) except for node 241, where it reaches 1.72:

![image](https://user-images.githubusercontent.com/1017210/131256771-7c333b4c-4cd4-419d-979c-01ffd27ffd04.png)

**Comment by avikivity** [comment]
Reads-with-misses is very low on most nodes, except on 241:

![image](https://user-images.githubusercontent.com/1017210/131256819-59105fac-629c-41c4-a955-1a80af493e9d.png)

**Comment by avikivity** [comment]
Reads-with-no-misses is consistent with reciprocal miss rate:

![image](https://user-images.githubusercontent.com/1017210/131256844-080c01ac-1761-41f6-910f-5f422e89d57d.png)

**Comment by avikivity** [comment]
The question is why did so many reads get sent to 241? With such low miss rates on the other nodes, very few reads should have been sent to 241, in order to keep the reads-with-misses the same on all nodes (which is the goal of heat based load balancing).

**Comment by gleb-cloudius** [comment]
Can you re-run with per table statistics enabled?

**Comment by bentsi** [comment]
how to enable "per table statistics"?

**Comment by nyh** [comment]
> how to enable "per table statistics"?

This is explained in docs/design-notes/metrics.md:

> To enable this flag and the per-table metrics, you can pass the parameters `--enable-keyspace-column-family-metrics 1` in the Scylla command line, or set this parameter in Scylla's configuration file.

**Comment by juliayakovlev** [comment]
> Can you re-run with per table statistics enabled?

Running. 
[Jenkins job](https://jenkins.scylladb.com/job/scylla-4.5/job/Reproducers/job/longevity-50gb-3days-table-stats/lastBuild/console)

[Monitor](http://13.48.177.95:3000/d/nCz7U0Vnk/longevity-50gb-3days-table-stats-scylla-per-server-metrics-nemesis-master?orgId=1&from=1630409707000&to=now&refresh=30s)

**Comment by nyh** [comment]
@slivne I see you assigned this issue to me, supposedly because it's somehow related to HWLB, but I don't understand at all how. People above suggested the problem is related to commit logs, to repair, to writes - how did all of this suddenly become a HWLB (i.e., **read**) issue?
I'll try to read the above comments to try to decipher how a write timeout might be is related to (read-side) HWLB, but if someone already understands the answer, it would be good to open a new issue instead of continuing to add (what appear to be) wild guesses to this one.

**Comment by nyh** [comment]
Ok, I have a better understanding now:

If this workload was read-only, @avikivity would have been right, something would seem very wrong with HWLB: Accoding to the numbers there, node 241 should do 1/10th (in the beginning) to 1/50th of the read work of the other nodes. And yet, it seems to be getting about 60% of the reads that other replicas get - between 18,000 and 19,000 of replica reads each second

But if I understand correctly, this workload is **not** read-only. It also has **materialized view writes**. And as you know, a MV has a read-before-write, and it is always local - regardless of HWLB... Could it be that the are all the reads we see? I can't figure out how many view writes - the total number of writes per second we have on this node is about 18,000-20,000 - are they all writes to a table that has a materialized view? how many materialized views? If so, they may explain almost all of the reads. @juliayakovlev can you please help me figure out how many writes we have per second to a table which has a materialized view?

The other possibility is that we have a serious bug in the HWLB algorithm which calculates the probabilities. I remember I had a unit test for the HWLB algorithm, but unfortunately it doesn't seem to be part of Scylla. I'll start to resurrect it now and test with the hit-ratio numbers that we have in this case. I _suspect_ that the problem is the MV reads, not HWLB, but I can't be sure before I test our HWLB algorithm on these numbers. It could also be a combination of both problems - MV and HWLB - somehow.

**Comment by juliayakovlev** [comment]
> Ok, I have a better understanding now:
> 
> If this workload was read-only, @avikivity would have been right, something would seem very wrong with HWLB: Accoding to the numbers there, node 241 should do 1/10th (in the beginning) to 1/50th of the read work of the other nodes. And yet, it seems to be getting about 60% of the reads that other replicas get - between 18,000 and 19,000 of replica reads each second
> 
> But if I understand correctly, this workload is **not** read-only. It also has **materialized view writes**. And as you know, a MV has a read-before-write, and it is always local - regardless of HWLB... Could it be that these are all (or most of) the reads we see? I can't figure out how many view writes - the total number of writes per second we have on this node is about 18,000-20,000 - are they all writes to a table that has a materialized view? how many materialized views? If so, they may explain almost all of the reads. @juliayakovlev can you please help me figure out how many writes we have per second to a table which has a materialized view?

@nyh 
I am not familiar with metrics with table name. I know about view metrics. If you know which metric to check I'll do it

> 
> The other possibility is that we have a serious bug in the HWLB algorithm which calculates the probabilities. I remember I had a unit test for the HWLB algorithm, but unfortunately it doesn't seem to be part of Scylla. I'll start to resurrect it now and test with the hit-ratio numbers that we have in this case. I _suspect_ that the problem is the MV reads, not HWLB, but I can't be sure before I test our HWLB algorithm on these numbers. It could also be a combination of both problems - MV and HWLB - somehow.

**Comment by slivne** [comment]
@nyh / @gleb-cloudius - there is a load running with per table metrics - so you can check it out

**Comment by slivne** [comment]
Running a load without MV's trying to see if this related

**Comment by slivne** [comment]
@gleb-cloudius ping

**Comment by gleb-cloudius** [comment]
The monitor is unusable. Unless you have hundreds of tables it should not be the case. How many tables you have?

**Comment by nyh** [comment]
Also, not being an expert in these tests, can you please help me decipher the workload (I saw the commands above, but don't know how to read them)? I know it is a mixture of reads and writes of some sort, but what percentage of requests are writes? Are *all* writes to a table with a materialized view or are there also tables without materialized view? Is there one materialized view per base table, or more than one?

**Comment by roydahan** [comment]
The monitor for a new run without MV:
http://3.250.109.220:3000/d/nvkJ-a47z/longevity-50gb-3days-no-mv-scylla-per-server-metrics-nemesis-master?orgId=1&from=now-24h&to=now&refresh=30s

**Comment by roydahan** [comment]
> The monitor is unusable. Unless you have hundreds of tables it should not be the case. How many tables do you have?

There are 2 tables.

**Comment by gleb-cloudius** [comment]
On Thu, Sep 02, 2021 at 04:45:49AM -0700, Roy Dahan wrote:
> > The monitor is unusable. Unless you have hundreds of tables it should not be the case. How many tables do you have?
> 
> There are 2 tables.
> 
Then I do not understand why the monitor is dead. 2 tables (together
with system once) would have added may be 100 counters. It should not
kill the monitor.

--
			Gleb.

**Comment by nyh** [comment]
> The monitor for a new run without MV:
> http://3.250.109.220:3000/d/nvkJ-a47z/longevity-50gb-3days-no-mv-scylla-per-server-metrics-nemesis-master?orgId=1&from=now-24h&to=now&refresh=30s
> (This is without per table statistics)

What do I see here? Was node 203 the one being rebuilt? 
It seems to me that node 203 is getting a lot less replica reads than the other nodes, as expected (right?), so we don't see the problem we had on the workload with the MV, right?

Something I don't understand is what happened around midnight - the number of replica reads on node 203 jumped almost instantly from 10k/sec to 40k/sec. The rate of cache misses decreased on 203 from 5k per second to almost zero, gradually, until the same time. This could be a HWLB problem - I'll think about it because I still can't figure out what kind of problem - but it's a different one from the one in the MV run, isn't it? It looks like here node 203 is getting too little work, not too much as in the MV workload?

**Comment by nyh** [comment]
A graph from the MV-less run:
![a](https://user-images.githubusercontent.com/584227/131847551-fa627bb9-5939-4fa2-80a9-eadd8c25bc18.jpg)

I think the explanation why 203 is getting so little work (in the new MV-less case) is #8815 which we didn't (yet?) backport to 4.5:

* Before #8815,  we had in the code the constant `max_hit_rate=0.999`. This means that all nodes besides 203 - nodes that have perfect cache hits (zero misses) - will all take hit_rate=0.999.
* Node 203 starts with hit_rate=0.2 but eventually goes up to almost perfect 1 as well. 
* Even when node 203 eventually gets to 0.989., its miss rate (0.011) is still 11 times more than that of 0.999 (0.001), so node 203 would get only 1/11th the number of reads as the other nodes. (note: we're not seeing that - we're seeing node 203 going up to about 1/6th of the other nodes. I don't know why).
* When node 203 reaches finally hit ratio = 0.989, there is a "phase shift": Another if() in `db/consistency_level.cc` kicks in - now all the hit rates - 0.999 and 0.989 are within `0.01` of another (absolute difference) - so the code switches off the HWLB calculations - and immediately instead of getting 1/11th of the requests as other nodes, node 203 gets the same number of requests. This is what we see in the graph - around midnight the number of reads per second on 203 makes a sharp jump from around 10k/sec to around 44k/sec and suddenly gets equal number of replica reads as the other nodes. (again, I don't know why the jump isn't 11-fold as I calculated above).

With the fix in #8815 , with max_hit_rate=0.95, the above numbers should change completely:
1. The hit_rate of the perfect nodes are pegged at 0.95.
2. When 203 reaches hit_rate 0.94, node 203 gets 83% (0.05/0.06) of the requests of the other nodes, and
3. At that point, the "0.01" optimization kicks in and all nodes get the same number of requests.
So the weird "phase shift" should be much smaller.

**Comment by nyh** [comment]
Please note that everything I said in the above comment doesn't explain the overloaded node problem we saw with the MV - quite the contrary - it explains an underloaded node problem (if it's even a problem...) that we saw in the MV-less run. 

@roydahan do you see any other or more serious problem on the MV-less run?

**Comment by gleb-cloudius** [comment]
On Thu, Sep 02, 2021 at 06:08:21AM -0700, nyh wrote:
> Please note that everything I said in the above comment doesn't explain the overloaded node problem we saw with the MV - quite the contrary - it explains an underloaded node problem (if it's even a problem...) that we saw in the MV-less run. 
> 
Can it be local reads that MV code does to build MV tables?

--
			Gleb.

**Comment by nyh** [comment]
> Can it be local reads that MV code does to build MV tables?

Good question. Above I suggested it might be read-before-write of the MV **writes** that this workload has, but a more careful examination of the number shows that these reads explain less than half of reads we see.
So if I understand you correctly, you're suggesting now that in addition to constantly getting MV writes, we also get **view** building because this nodes is streaming in sstables and then view-building them? @roydahan does this workload really do that (stream data into the slow node)?

If this is the case, then indeed most of the reads we're counting are probably (?) caused by view writes or builds - and have nothing to do with HWLB which probably works as expected and makes sure that very few user reads are serviced by this replica - but it can't protect against the view-write-initiated reads!

If this is really the explanation, then I wonder why this is a regression in 4.5: The new node is **justly** more loaded than others, because of all those view read-before-writes that it does that are both more numerous (because it does view building) and slower (colder cache) than what the other nodes do. The only question is how Scylla handles this overload of read operations and why it regressed in 4.5.

I have a wild guess: Could it be that 4.5 changed something in the "read concurency semaphore"? When Scylla has a lot of view-building and read-before-write activity, maybe this for some reason gets all the read permits, and the user reads don't get enough permits or whatever, and time out?

Do we also see this problem on master, or just on 4.5?

**Comment by gleb-cloudius** [comment]
On Thu, Sep 02, 2021 at 06:35:56AM -0700, nyh wrote:
> > Can it be local reads that MV code does to build MV tables?
> 
> Good question. Above I suggested it might be read-before-write of the MV **writes** that this workload has, but a more careful examination of the number shows that these reads explain less than half of reads we see.
> So if I understand you correctly, you're suggesting now that in addition to constantly getting MV writes, we also get **view** building because this nodes is streaming in sstables and then view-building them? @roydahan does this workload really do that (stream data into the slow node)?
> 
As far as I understand the node is getting rebuilt.

--
			Gleb.

**Comment by aleksbykov** [comment]
Issue was reproduced with next scylla-version:
Kernel version: `5.4.0-1035-aws`
Scylla version (or git commit hash): `4.5.rc6-0.20210817.88314fedf`
Cluster size: 6 nodes (i3.4xlarge)
Scylla running with shards number (live nodes):
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-1 (16.170.172.168 | 10.0.2.145): 14 shards
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-2 (16.170.159.226 | 10.0.3.141): 14 shards
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-3 (13.48.130.253 | 10.0.3.30): 14 shards
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-4 (16.170.98.191 | 10.0.2.115): 14 shards
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-5 (13.48.70.50 | 10.0.0.210): 14 shards
   longevity-tls-50gb-3d-longevit-db-node-0a30abb8-6 (13.53.36.96 | 10.0.0.62): 14 shards
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-074912d764ff480e5` (aws: eu-north-1)

Test: `longevity-50gb-3days-table-stats-1nemesis`
Test name: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):
- [longevity-50GB-3days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/ff5a8aba31edf53da4ac31b2bfd10b9f82ab8af0/test-cases/longevity/longevity-50GB-3days-authorization-and-tls-ssl.yaml)
Logs:
grafana - [https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004110/grafana-screenshot-overview-20210903_004110-longevity-tls-50gb-3d-longevit-monitor-node-0a30abb8-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004110/grafana-screenshot-overview-20210903_004110-longevity-tls-50gb-3d-longevit-monitor-node-0a30abb8-1.png)
db-cluster - [https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/db-cluster-0a30abb8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/db-cluster-0a30abb8.tar.gz)
loader-set - [https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/loader-set-0a30abb8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/loader-set-0a30abb8.tar.gz)
monitor-set - [https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/monitor-set-0a30abb8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/monitor-set-0a30abb8.tar.gz)
sct-runner - [https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/sct-runner-0a30abb8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0a30abb8-4f41-4c32-a350-3b3d6305de71/20210903_004936/sct-runner-0a30abb8.tar.gz)
[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-4.5/job/Reproducers/job/longevity-50gb-3days-table-stats-1nemesis/2/)

Cluster is available:
```
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-1      | eu-north-1a | 16.170.172.168 | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-2      | eu-north-1a | 16.170.159.226 | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-3      | eu-north-1a | 13.48.130.253  | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-4      | eu-north-1a | 16.170.98.191  | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-5      | eu-north-1a | 13.48.70.50    | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |
| longevity-tls-50gb-3d-longevit-db-node-0a30abb8-6      | eu-north-1a | 13.53.36.96    | 0a30abb8-4f41-4c32-a350-3b3d6305de71 |

```

Monitoring of 4.5.rc6: http://13.51.238.56:3000/d/YsClRE47k/4-5-rc6-1nemesis-table-stats-longevity-50gb-3days-scylla-per-server-metrics-nemesis-master?orgId=1&from=1630586480000&to=now&refresh=30s&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=CassandraStressEvent&var-sct_tags=InfoEvent

Job was running with next parameters;
- enable-keyspace-column-family-metrics
- instance_type_monitor: 'c5.4xlarge'
- nemesis_class_name: 'RebuildStreamingErrMonkey'

Job failed after 12hours. According to to metrics issue #9053. Job was terminated due critical error:
Java.io.IOException: Operation x10 on key(s) [5030394d4d36504b3330]: Error executing: (WriteTimeoutException): Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
Write & read timeouts 
![Screenshot from 2021-09-03 15-06-54](https://user-images.githubusercontent.com/44991177/131977619-4896fd4d-f221-4169-8aa8-9e59b61cf147.png)
 Replica Write timeoutes
![Screenshot from 2021-09-03 15-06-38](https://user-images.githubusercontent.com/44991177/131977632-322a3e1b-73a4-49e3-a205-5d2c00460d4c.png)

Same tests with same configuration but with scylla 4.4 is still running.
Jenkins job:
https://jenkins.scylladb.com/job/scylla-4.5/job/Reproducers/job/longevity-50gb-3days-4.4.4-table-stats-1nemesis/3/console

Monitor:
http://13.48.106.148:3000/d/whD2iEVnz/4-4-4-table-stats-1nemesis-longevity-50gb-3days-scylla-per-server-metrics-nemesis-master?orgId=1&from=1630588410000&to=now&refresh=30s
```
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-1      | eu-north-1a | 13.49.78.127  | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-2      | eu-north-1a | 13.53.35.227  | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-3      | eu-north-1a | 13.51.161.125 | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-4      | eu-north-1a | 13.48.133.80  | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-5      | eu-north-1a | 13.49.46.168  | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
| longevity-tls-50gb-3d-longevit-db-node-8064bd79-6      | eu-north-1a | 13.49.70.85   | 8064bd79-74eb-4918-90b6-fcb8605dc4e7 
```

**Comment by nyh** [comment]
I haven't yet looked at the latest monitoring results, but just want to note that I just sent to the mailing list a patch with a new unit test for HWLB's calculations, which makes it unlikely (though not impossible) that HWLB is to blame for the overloading of the cold-cache node. The test confirms that HWLB in all cases tested results in the correct distribution of requests, and wouldn't have caused a node that get 1/10th of the requests to get 1/2 instead. The new test reproduces another bug (#9285) but it can't explain what we see in this test.

**Comment by slivne** [comment]
highlighted part 

4.5.rc6
http://13.51.238.56:3000/d/detailed-4-5/detailed?orgId=1&from=1630605222669&to=1630643677456

please check the hwlb metrics look the same 

4.4.4 metrics please look at them as well - it seems they also have places "stuck".

With regards to MV I think the following commit part of 4.4.4 and 4.5 maybe relevant 

commit c66661c5824fe075fd0729fb09818d849dfc4822
Author: Piotr Sarna <sarna@scylladb.com>
Date:   Sun Apr 26 12:53:35 2020 +0200

    table: bypass cache when generating view updates from streaming
    
    There's no indication that data needed for generating view updates
    from staging sstables is going to be immediately useful for the
    user, and a large amount of it can push hot rows out of the cache,
    thus deteriorating performance.
    
    Fixes #6233
    Tests: unit(dev)

So read before writes to mv should be bypassing cache and if they are accounting for the majority of the reads.

Did we do any change in mv update code in 4.5 not part of 4.4 - checking if something could have caused another change.

**Comment by roydahan** [comment]
The issue has reproduced on 4.4.4 as well.
It took 48h but eventually reproduced.

Monitor:
http://13.48.106.148:3000/d/whD2iEVnz/4-4-4-table-stats-1nemesis-longevity-50gb-3days-scylla-per-server-metrics-nemesis-master?orgId=1&from=1630756395763&to=1630772552557

detailed dashboard at the time of the issue: http://13.48.106.148:3000/d/detailed-4-4/detailed?orgId=1&from=1630757352987&to=1630765126813

Live screenshots:
https://snapshot.raintank.io/dashboard/snapshot/lfxol4oowwPKBZ7XZpkRAsYEUYZ7OIxL
https://snapshot.raintank.io/dashboard/snapshot/bok52VtbS39CSyG4ZJp6oORZwF5qAdXq

**Comment by slivne** [comment]
As it reproduced on 4.4.4 we do not consider this a blocker and will continue to investigate - while this doesn't block 4.5 for now

**Comment by nyh** [comment]
> highlighted part
> 
> 4.5.rc6
> http://13.51.238.56:3000/d/detailed-4-5/detailed?orgId=1&from=1630605222669&to=1630643677456
> 
> please check the hwlb metrics look the same

I don't understand what I'm seeing there at all. The "Reciprocal Miss Rate" graph shows wildly different lines for the different nodes, 2 or 3 nodes show almost perfect cache (rmr = 100 - @amnonh can correct me but I think it is cut off at 100 completely arbitrarily, and may actually be 1000) while 3 or 4 others have very low - rmr as low as 2 indicates just a 50% cache hit rate. As a result we see "Read per instance - Replica" at a very different place between nodes (while the client is obviously sending all the coordinators the same number of requests - request served per instance is the same.

So if we're back to thinking that HWLB has something to do with the problem, it seems that its problem is not the algorithm (which apparently does what's expected from it) but rather the inputs to the algorithm - the hit rates (or the "reciprocal miss rate") which seem really weird. If we assume that only one node was rebooted, why do we see very low cache hit rates for additional nodes?

I wonder if the patch that @slivne found is indeed relevant:

> commit [c66661c](https://github.com/scylladb/scylla/commit/c66661c5824fe075fd0729fb09818d849dfc4822)
> Author: Piotr Sarna [sarna@scylladb.com](mailto:sarna@scylladb.com)
> Date: Sun Apr 26 12:53:35 2020 +0200
> 
> ```
> table: bypass cache when generating view updates from streaming

Could it be that reads that deliberate pass the cache are counted as cache misses? I'll have to try follow the code and see how exactly the cache misses are counted.

What's strange, though, is why two nodes out of the 6 retained their perfect-looking cache hit ratio. Could it be that two nodes were not involved in streaming? How?

**Comment by roydahan** [comment]
> > highlighted part
> > 4.5.rc6
> > http://13.51.238.56:3000/d/detailed-4-5/detailed?orgId=1&from=1630605222669&to=1630643677456
> > please check the hwlb metrics look the same
> 
> I don't understand what I'm seeing there at all. The "Reciprocal Miss Rate" graph shows wildly different lines for the different nodes, 2 or 3 nodes show almost perfect cache (rmr = 100 - @amnonh can correct me but I think it is cut off at 100 completely arbitrarily, and may actually be 1000) while 3 or 4 others have very low - rmr as low as 2 indicates just a 50% cache hit rate. As a result we see "Read per instance - Replica" at a very different place between nodes (while the client is obviously sending all the coordinators the same number of requests - request served per instance is the same.
> 
> So if we're back to thinking that HWLB has something to do with the problem, it seems that its problem is not the algorithm (which apparently does what's expected from it) but rather the inputs to the algorithm - the hit rates (or the "reciprocal miss rate") which seem really weird. If we assume that only one node was rebooted, why do we see very low cache hit rates for additional nodes?

This test is a run of  the same nemesis over and over till the issue has reproduced.
Every time the nemesis run it can pick a different node.
So, in this case it make sense that some of the other nodes has a cold cache because they may have rebooted few minutes earlier by the nemesis.

**Comment by nyh** [comment]
> This test is a run of the same nemesis over and over till the issue has reproduced.
> Every time the nemesis run it can pick a different node.
> So, in this case it make sense that some of the other nodes has a cold cache because they may have rebooted few minutes earlier by the nemesis.

Are you saying that in the montitor sent by Shlomi, http://13.51.238.56:3000/d/detailed-4-5/detailed?orgId=1&from=1630605222669&to=1630643677456 - spanning about 7 hours - there were actually multiple nodes being rebooted and rebuilt?  Could it be that the same node is doing this multiple times? E.g., node 210 seems to begin the period that Shlomi highlighted with a cold cache, and then the cache fills up gradually, until four hours later it has almost-perfect cache (rmr=75, corresponds to cache hit rate of %98.6), at which points the cache becomes crappy again - was this node restarted again? Or the answer is no, maybe it started to get involved in streaming which is wrongly counted as cache misses?

Can I see somewhere int he log which node was started at which time?

**Comment by roydahan** [comment]
> Are you saying that in the montitor sent by Shlomi, http://13.51.238.56:3000/d/detailed-4-5/detailed?orgId=1&from=1630605222669&to=1630643677456 - spanning about 7 hours - there were actually multiple nodes being rebooted and rebuilt? Could it be that the same node is doing this multiple times? E.g., node 210 seems to begin the period that Shlomi highlighted with a cold cache, and then the cache fills up gradually, until four hours later it has almost-perfect cache (rmr=75, corresponds to cache hit rate of %98.6), at which points the cache becomes crappy again - was this node restarted again? Or the answer is no, maybe it started to get involved in streaming which is wrongly counted as cache misses?
> 
Yes, any node can be restarted in each nemesis, they are picked randomly.

> Can I see somewhere in the monitor which node was started at which time?
Yes, if you turn on “Normal events” you’ll see grafana annotations at each point a node restarted.

**Comment by slivne** [comment]
@nyh ping

**Comment by fgelcer** [comment]
i'm experiencing this issue on master.. at least 3 different longevity jobs failed with this issue.
in one of them, we had `NoCorruptRepair` running 3 times, and failing all of them.

here the info about the 1st job:
*Installation details*
Kernel version: `5.4.0-1035-aws`
Scylla version (or git commit hash): `4.6.dev-0.20211017.1d751491a with build-id 1ab3a33364ab7aff5e76f592f4a3e18d2b36c98f`
Cluster size: 6 nodes (i3.4xlarge)
Scylla running with shards number (live nodes):
   longevity-tls-50gb-3d-master-db-node-14ef667a-1 (13.49.65.150 | 10.0.1.209): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-3 (13.51.193.26 | 10.0.3.159): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-5 (13.53.101.85 | 10.0.1.234): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-6 (13.48.106.145 | 10.0.3.183): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-7 (13.53.87.111 | 10.0.1.125): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-8 (13.51.178.182 | 10.0.1.6): 14 shards
Scylla running with shards number (terminated nodes):
   longevity-tls-50gb-3d-master-db-node-14ef667a-2 (16.170.143.175 | 10.0.1.241): 14 shards
   longevity-tls-50gb-3d-master-db-node-14ef667a-4 (13.48.104.40 | 10.0.2.222): 14 shards
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-042d6f0aeb4087705` (aws: eu-north-1)


Test: `longevity-50gb-3days`
Test name: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):
- [longevity-50GB-3days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/27b59253648831cba683ba58818a8fb4791a5b64/test-cases/longevity/longevity-50GB-3days-authorization-and-tls-ssl.yaml)

Restore Monitor Stack command: `$ hydra investigate show-monitor 14ef667a-b0a4-437a-a521-1c6fe6443936`
Show all stored logs command: `$ hydra investigate show-logs 14ef667a-b0a4-437a-a521-1c6fe6443936`

Test id: `14ef667a-b0a4-437a-a521-1c6fe6443936`


Logs:
grafana - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_102222/grafana-screenshot-longevity-50gb-3days-scylla-per-server-metrics-nemesis-20211019_102511-longevity-tls-50gb-3d-master-monitor-node-14ef667a-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_102222/grafana-screenshot-longevity-50gb-3days-scylla-per-server-metrics-nemesis-20211019_102511-longevity-tls-50gb-3d-master-monitor-node-14ef667a-1.png)
grafana - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_102222/grafana-screenshot-overview-20211019_102223-longevity-tls-50gb-3d-master-monitor-node-14ef667a-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_102222/grafana-screenshot-overview-20211019_102223-longevity-tls-50gb-3d-master-monitor-node-14ef667a-1.png)
db-cluster - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/db-cluster-14ef667a.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/db-cluster-14ef667a.tar.gz)
loader-set - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/loader-set-14ef667a.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/loader-set-14ef667a.tar.gz)
monitor-set - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/monitor-set-14ef667a.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/monitor-set-14ef667a.tar.gz)
sct-runner - [https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/sct-runner-14ef667a.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/14ef667a-b0a4-437a-a521-1c6fe6443936/20211019_103543/sct-runner-14ef667a.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-50gb-3days/103/)

**Comment by slivne** [comment]
moving to 4.6 - we disabled commitlog hard limit

**Comment by slivne** [comment]
4.6 was branched and commit log hard limit is not enforced

**Comment by elcallio** [comment]
Is this still valid?

**Comment by fgelcer** [comment]
> @fgelcer have this issue seen again since we disabled hard limit (any version greater than 4.6 including master).

checking last reported issues for 5.0 and 2022.1 this one wasn't reported (are there other similar issues, that perhaps people used to report?)

**Comment by DoronArazii** [comment]
AFAIK commitlog-hard-limit planned for 5.3

Pushing it one milestone forward.

**Comment by DoronArazii** [comment]
@eliransin can we close it?

**Comment by mykaul** [comment]
Closing - commit log hard limit is in.
